File: backend/src/main/java/com/ayigroup/rabbitmq/playground/scenario/ScenarioDefaults.java
package com.ayigroup.rabbitmq.playground.scenario;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Configuración de demostración inicial para cada tipo de Exchange.
 * Estos valores son solo el punto de partida: el usuario puede
 * modificarlos desde la UI una vez creado el escenario.
 */
public final class ScenarioDefaults {

    private ScenarioDefaults() {
    }

    public static List<QueueConfig> build(ExchangeType type) {
        return switch (type) {
            case FANOUT -> List.of(
                    QueueConfig.builder().label("Cola A").build(),
                    QueueConfig.builder().label("Cola B").build(),
                    QueueConfig.builder().label("Cola C").build()
            );
            case DIRECT -> List.of(
                    QueueConfig.builder().label("Errores criticos").bindingKey("error").build(),
                    QueueConfig.builder().label("Logs info").bindingKey("info").build(),
                    QueueConfig.builder().label("Warnings").bindingKey("warning").build()
            );
            case TOPIC -> List.of(
                    QueueConfig.builder().label("Europa").pattern("eu.#").build(),
                    QueueConfig.builder().label("Clima global").pattern("#.temperatura").build(),
                    QueueConfig.builder().label("Estados Unidos").pattern("us.#").build()
            );
            case HEADERS -> List.of(
                    QueueConfig.builder()
                            .label("Facturas en espanol")
                            .xMatch("all")
                            .headers(mapOf("type", "invoice", "lang", "es"))
                            .build(),
                    QueueConfig.builder()
                            .label("Reportes en ingles")
                            .xMatch("all")
                            .headers(mapOf("type", "report", "lang", "en"))
                            .build(),
                    QueueConfig.builder()
                            .label("Facturas procesadas (any)")
                            .xMatch("any")
                            .headers(mapOf("type", "invoice", "status", "processed"))
                            .build()
            );
            case DEFAULT -> List.of(
                    QueueConfig.builder().label("pedidos").build(),
                    QueueConfig.builder().label("notificaciones").build()
            );
            case EXCHANGE_TO_EXCHANGE -> List.of(
                    QueueConfig.builder()
                            .label("Urgentes")
                            .pattern("pedido.urgente.#")
                            .boundExchange(BoundExchange.PRIMARY)
                            .build(),
                    QueueConfig.builder()
                            .label("Todos los pedidos")
                            .pattern("pedido.#")
                            .boundExchange(BoundExchange.SECONDARY)
                            .build(),
                    QueueConfig.builder()
                            .label("Cancelaciones")
                            .pattern("pedido.cancelado.*")
                            .boundExchange(BoundExchange.SECONDARY)
                            .build()
            );
            case ALTERNATE_EXCHANGE -> List.of(
                    QueueConfig.builder()
                            .label("Urgentes")
                            .bindingKey("urgente")
                            .boundExchange(BoundExchange.PRIMARY)
                            .build(),
                    QueueConfig.builder()
                            .label("Normales")
                            .bindingKey("normal")
                            .boundExchange(BoundExchange.PRIMARY)
                            .build(),
                    QueueConfig.builder()
                            .label("Huerfanos")
                            .boundExchange(BoundExchange.SECONDARY)
                            .build()
            );
        };
    }

    /** Patrón por defecto del puente Exchange 1 -> Exchange 2, solo para EXCHANGE_TO_EXCHANGE. */
    public static String defaultBridgeBindingKey() {
        return "pedido.#";
    }

    private static Map<String, String> mapOf(String... kv) {
        Map<String, String> map = new LinkedHashMap<>();
        for (int i = 0; i < kv.length; i += 2) {
            map.put(kv[i], kv[i + 1]);
        }
        return map;
    }

    public static List<QueueConfig> copy(List<QueueConfig> queues) {
        List<QueueConfig> result = new ArrayList<>();
        for (QueueConfig q : queues) {
            result.add(q.copy());
        }
        return result;
    }
}

File: src/lib/engine/scenarioDefaults.ts
import type { ExchangeType, QueueConfig } from "../../types";

/** Port 1:1 de `ScenarioDefaults` (backend Java): la topología de demo por cada uno de los 8 tipos. */

function q(partial: Partial<QueueConfig> & { label: string }): QueueConfig {
  return {
    name: "", // se completa en scenarioStore al crear el escenario (necesita sessionId/typeSlug)
    bindingKey: null,
    pattern: null,
    headers: null,
    xMatch: null,
    ackMode: "AUTO",
    boundExchange: "PRIMARY",
    ...partial,
  };
}

export function buildDefaultQueues(type: ExchangeType): QueueConfig[] {
  switch (type) {
    case "FANOUT":
      return [q({ label: "Cola A" }), q({ label: "Cola B" }), q({ label: "Cola C" })];

    case "DIRECT":
      return [
        q({ label: "Errores criticos", bindingKey: "error" }),
        q({ label: "Logs info", bindingKey: "info" }),
        q({ label: "Warnings", bindingKey: "warning" }),
      ];

    case "TOPIC":
      return [
        q({ label: "Europa", pattern: "eu.#" }),
        q({ label: "Clima global", pattern: "#.temperatura" }),
        q({ label: "Estados Unidos", pattern: "us.#" }),
      ];

    case "HEADERS":
      return [
        q({ label: "Facturas en espanol", xMatch: "all", headers: { type: "invoice", lang: "es" } }),
        q({ label: "Reportes en ingles", xMatch: "all", headers: { type: "report", lang: "en" } }),
        q({ label: "Facturas procesadas (any)", xMatch: "any", headers: { type: "invoice", status: "processed" } }),
      ];

    case "DEFAULT":
      return [q({ label: "pedidos" }), q({ label: "notificaciones" })];

    case "EXCHANGE_TO_EXCHANGE":
      return [
        q({ label: "Urgentes", pattern: "pedido.urgente.#", boundExchange: "PRIMARY" }),
        q({ label: "Todos los pedidos", pattern: "pedido.#", boundExchange: "SECONDARY" }),
        q({ label: "Cancelaciones", pattern: "pedido.cancelado.*", boundExchange: "SECONDARY" }),
      ];

    case "ALTERNATE_EXCHANGE":
      return [
        q({ label: "Urgentes", bindingKey: "urgente", boundExchange: "PRIMARY" }),
        q({ label: "Normales", bindingKey: "normal", boundExchange: "PRIMARY" }),
        q({ label: "Huerfanos", boundExchange: "SECONDARY" }),
      ];

    case "DEAD_LETTER_EXCHANGE":
      return [
        q({ label: "Procesar pago", bindingKey: "pago.nuevo", boundExchange: "PRIMARY" }),
        q({ label: "Pagos fallidos", boundExchange: "SECONDARY" }),
      ];

    default:
      throw new Error(`Tipo de escenario desconocido: ${type}`);
  }
}

export const DEFAULT_BRIDGE_BINDING_KEY = "pedido.#";
